In [4]:
import pandas as pd
import numpy as np
import sys
from pytrends.request import TrendReq
import time
import warnings
warnings.filterwarnings("ignore")
# sys.path.append('../methods')
# from SVI_methods import index_with_dates

C:\Users\x4287225\AppData\Local\Temp\ipykernel_18824\2519467636.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [5]:


def index_with_dates(keywords, start_date, end_date):
     """
     keywords: list of keywords => ['AAPL']
     start_date: a date as type string => '2018-01-01'
     end_date: a date as type string => '2018-01-01'
     when time interval is under than 8 months it returns on a daily basis
     when time interval is under than 6 years it returns on a weekly basis
     when time interval more than 6 years it returns on a yearly basis
     # cat='' Financial Markets: 1163
     """
     try:
          pytrends = TrendReq(hl='en-US', tz=360, requests_args={'verify':False})
          timeframe = start_date + " " + end_date
          pytrends.build_payload(kw_list=keywords, timeframe=timeframe, geo='', gprop='')
          results = pytrends.interest_over_time()
          return results
     except Exception as e:
          print(e)
          print(f'error has occured for {keywords}, {start_date} to {end_date}')

### Monthly SVI Extraction

In [6]:
companies = pd.read_csv("../data/companies_after_2005.csv")
symbols = np.array(companies['Symbol'])

In [28]:
# monthly_svi = pd.DataFrame(columns = ['symbol','date', 'svi'])

bad_symbols = ['ADBE', 'ACN', 'PEP']
empty_symbols = []
for s in symbols[87:]: 
     time.sleep(6)
     x = index_with_dates([s], "2005-01-01", "2023-12-01")
     if isinstance(x, pd.DataFrame):
          if(len(x) != 0):
               new_df = pd.DataFrame(columns = ['symbol','date', 'svi'])
               new_df['date'] = x.index
               new_df['svi'] = x.iloc[:,0].values
               new_df['symbol'] = s
               monthly_svi = pd.concat([monthly_svi,new_df], axis = 0)
               monthly_svi.reset_index(inplace=True, drop=True)
          else:
               print(s, "is an empty symbol")
               empty_symbols.append(s)
     else:
          print(s, "has an error")
          bad_symbols.append(s)


6861.T is an empty symbol
8035.T is an empty symbol
601628.SS is an empty symbol
AI.PA is an empty symbol
601318.SS is an empty symbol
The request failed: Google returned a response with code 429
error has occured for ['ADI'], 2005-01-01 to 2023-12-01
ADI has an error
ABBN.SW is an empty symbol
4063.T is an empty symbol
SBIN.NS is an empty symbol
BHARTIARTL.NS is an empty symbol
ATCO-B.ST is an empty symbol
000660.KS is an empty symbol
6501.T is an empty symbol
INVE-B.ST is an empty symbol
ZURN.SW is an empty symbol
The request failed: Google returned a response with code 429
error has occured for ['MRK.DE'], 2005-01-01 to 2023-12-01
MRK.DE has an error
The request failed: Google returned a response with code 429
error has occured for ['IBE.MC'], 2005-01-01 to 2023-12-01
IBE.MC has an error
002594.SZ is an empty symbol
The request failed: Google returned a response with code 429
error has occured for ['PH'], 2005-01-01 to 2023-12-01
PH has an error
The request failed: Google returned a

In [32]:
very_bad_symbols = []
# empty_symbols = []
for s in bad_symbols: 
     time.sleep(20)
     x = index_with_dates([s], "2005-01-01", "2023-12-01")
     if isinstance(x, pd.DataFrame):
          if(len(x) != 0):
               new_df = pd.DataFrame(columns = ['symbol','date', 'svi'])
               new_df['date'] = x.index
               new_df['svi'] = x.iloc[:,0].values
               new_df['symbol'] = s
               monthly_svi = pd.concat([monthly_svi,new_df], axis = 0)
               monthly_svi.reset_index(inplace=True, drop=True)
          else:
               print(s, "is an empty symbol")
               empty_symbols.append(s)
     else:
          print(s, "has an error")
          very_bad_symbols.append(s)

MRK.DE is an empty symbol
IBE.MC is an empty symbol
601899.SS is an empty symbol
NDA-FI.HE is an empty symbol
DB1.DE is an empty symbol
600276.SS is an empty symbol
601998.SS is an empty symbol


In [34]:
monthly_svi['symbol'].nunique()

392

In [36]:
monthly_svi.to_csv("../data/monthly_svi.csv")
np.save("../data/empty_symbols.npy", empty_symbols)

### Weekly SVI

* We have 19 years, the divide will be 2005/1 to 2009/12, 2010/1 to 2014/12, 2015/1 to 2019/12, 2020/1 to 2023/12 . Th reasons is because of the fact that we need weekly data and the time intervals should be less than 6 years.
* We have 392 symbols


In [61]:
monthly_svi['symbol'].unique()

array(['MSFT', 'AAPL', 'NVDA', 'AMZN', 'GOOG', 'BRK-B', 'LLY', 'TSM',
       'NVO', 'JPM', 'UNH', 'WMT', 'XOM', 'JNJ', 'PG', 'ASML', 'HD',
       'TCEHY', 'MRK', 'COST', 'TM', 'ORCL', 'CVX', 'AMD', 'CRM', 'BAC',
       'KO', 'NFLX', 'TMO', 'MCD', 'SAP', 'DIS', '601857.SS', 'AZN',
       'ABT', 'CSCO', 'TMUS', 'WFC', 'DHR', 'INTC', 'INTU', 'IBM', 'QCOM',
       'VZ', 'AMAT', 'CMCSA', 'CAT', 'NKE', 'PFE', 'AXP', 'AMGN', 'UNP',
       'TXN', 'SIE.DE', 'MS', 'RY', 'HDB', 'ISRG', 'SPGI', 'SYK', 'IDEXY',
       'LOW', 'COP', 'HON', 'BKNG', 'UL', 'UPS', 'PLD', 'AIR.PA', 'GS',
       'BA', 'BUD', 'LRCX', 'T', 'RTX', 'DTE.DE', 'BLK', 'SCHW', 'NEE',
       'SNY', 'PBR', 'MDT', 'RIO', 'TJX', 'PGR', 'SONY', 'VRTX', 'TD',
       'SBUX', 'C', 'ALV.DE', 'REGN', 'ADP', 'LMT', 'DE', 'BMY', 'BP',
       'CI', 'MMC', 'CVS', 'MDLZ', 'BSX', 'EL.PA', 'CSL.AX', 'KLAC',
       'GILD', 'FI', 'MU', 'AMT', 'IBN', '9983.T', 'GSK', 'INFY', 'SNPS',
       '8058.T', 'CNI', '0883.HK', 'HCA', 'DEO', '9984.T', 'WM', 'R

In [74]:
start_intervals = ["2005-01-01", "2010-01-01", "2015-01-01", "2020-01-01"]
end_intervals = ["2009-12-31", "2014-12-31", "2019-12-31", "2023-12-31"]
weekly_symbols = monthly_svi['symbol'].unique()
event_log = pd.DataFrame(columns=['symbol','start_date', 'end_date'])
for s in weekly_symbols:
     for d in range(4):
          event = [s, start_intervals[d],end_intervals[d]]
          event_log.loc[len(event_log)] = event
          

event_log


symbol  start_date    end_date
0      MSFT  2005-01-01  2009-12-31
1      MSFT  2010-01-01  2014-12-31
2      MSFT  2015-01-01  2019-12-31
3      MSFT  2020-01-01  2023-12-31
4      AAPL  2005-01-01  2009-12-31
...     ...         ...         ...
1563    HUM  2020-01-01  2023-12-31
1564   GIII  2005-01-01  2009-12-31
1565   GIII  2010-01-01  2014-12-31
1566   GIII  2015-01-01  2019-12-31
1567   GIII  2020-01-01  2023-12-31

[1568 rows x 3 columns]

In [70]:
Weekly_svi = pd.DataFrame(columns = ['symbol','date', 'svi'])
error_log = pd.DataFrame(columns=['symbol','start_date', 'end_date', 'error'])

for r in range(len(event_log)):
     time.sleep(6)
     x = index_with_dates([event_log['symbol'][r]], event_log['start_date'][r], event_log['end_date'][r])
     if isinstance(x, pd.DataFrame):
          if(len(x) != 0):
               new_df = pd.DataFrame(columns = ['symbol','date', 'svi'])
               new_df['date'] = x.index
               new_df['svi'] = x.iloc[:,0].values
               new_df['symbol'] = event_log['symbol'][r]
               Weekly_svi = pd.concat([Weekly_svi,new_df], axis = 0)
               Weekly_svi.reset_index(inplace=True, drop=True)
          else:
               print(event_log['symbol'][r], "is an empty symbol")
               error_log.loc[len(error_log)] = [event_log['symbol'][r], event_log['start_date'][r], event_log['end_date'][r], "empty symbol"]
     else:
          print(event_log['symbol'][r], "has an error")
          error_log.loc[len(error_log)] = [event_log['symbol'][r], event_log['start_date'][r], event_log['end_date'][r], "server error"]

In [64]:
Weekly_svi.to_csv("../data/weekly_svi.csv")

In [71]:
Weekly_svi['symbol'].unique()

array(['MSFT', 'AAPL', 'NVDA'], dtype=object)

In [69]:
event_log

symbol  start_date    end_date
0      MSFT  2005-01-01  2009-12-31
1      MSFT  2010-01-01  2014-12-31
2      MSFT  2015-01-01  2019-12-31
3      MSFT  2020-01-01  2023-12-31
4      AAPL  2005-01-01  2009-12-31
...     ...         ...         ...
1563    HUM  2020-01-01  2023-12-31
1564   GIII  2005-01-01  2009-12-31
1565   GIII  2010-01-01  2014-12-31
1566   GIII  2015-01-01  2019-12-31
1567   GIII  2020-01-01  2023-12-31

[1568 rows x 3 columns]

In [73]:
error_log

Empty DataFrame
Columns: [symbol, start_date, end_date, error]
Index: []